### DISCUSSION TITLE
Welcome to ControlSystemIdentification.jl Discussions!

### DISCUSSION THREAD

**Discussion by baggepinnen** [baggepinnen/ControlSystemIdentification.jl#51] [General]
Feel free to ask usage questions or questions about system identification in general.

**Comment by DionysisPap** [comment]
Hi, i wanted to ask what i am doing wrong when i am trying to add this specific package to VS Code.
I have the last version of Julia v1.7.3 and i have the extension to VS Code.
But when i go to add the package i have an error that say "9 dependencies errored". And the errors are about:

DSP
MonteCarloMeasurements
DiffEqBase
LowLevelParticleFilters
DiffEqCallbacks
OrdinaryDiffEq
DelayDiffEq
ControlSystems
ControlSystemIdentification

I dont know if i have to add the specific package with some alternative way. Thanks

**Comment by DionysisPap** [comment]
I solve it by adding the version 2.3.0, that was successfully compliled and then updating to the newer version 2.4.0. I can't figure out what was the problem exactly.

**Reply by baggepinnen** [reply]
Hello, I'm happy you solved the problem!

**Comment by m-bossart** [comment]
Hello,

I'm looking to perform System ID on data which comes from multiple independent experiments of the same system. Is there a way to deal with this type of data and fit a model to all of the data. Something similar to this functionality in MATLAB:
https://ch.mathworks.com/help/ident/ug/dealing-with-multi-experiment-data-and-merging-models.html
Thanks!

**Reply by baggepinnen** [reply]
Hey!

We currently have no good way to handle this 😏The only option right now would be to use lower-level functions and manually assemble multi-experiment regressor matrices using, e.g., `getARXregressor`

**Reply by m-bossart** [reply]
Thanks for the quick reply! I am not very familiar with System ID--can you share any resources regarding how to manually combine the regressor matrices. Is there a standard way to derive these merged models?

**Reply by baggepinnen** [reply]
What method to use would depend on your situation,
- Do all experiments share the same initial condition, or is the initial condition always 0?
- How many inputs and outputs does the system have?

**Reply by baggepinnen** [reply]
For example, the main part ot the `arx` function looks like this
```julia
    y_train, A = getARXregressor(y, u, na, nb; inputdelay)
    w = ls(A, y_train, λ, estimator) |> vec
    a, b = params2poly2(w, na, nb; inputdelay)
    model = tf(b, a, h)
```
To use multiple datasets, you would just create several different `y_train, A`, and then concatenate them like
```
Y_train = [y_ytrain[1]; y_train[2], ...]
A = [A[1]; A[2]; ...]
```
and then use these concatenated matrices when calling `ls`. `arx` is not always applicable, and even when it is, it's not always the best estimation method, it all comes down to the properties of your system and experiments.

**Reply by baggepinnen** [reply]
Here's a version of ERA/OKID that uses several experiments
```julia
using ControlSystemIdentification, ControlSystemsBase, Statistics
Ts = 0.1
G = c2d(tf(1, [1,1,1]), Ts)

# Create several "experiments"
ds = map(1:5) do i
    u = randn(1, 1000)
    y, t, x = lsim(G, u)
    yn = y + 0.2randn(size(y))
    iddata(yn, u, Ts)
end

Ys = okid.(ds, 2, round(Int, 10/Ts)) # Estimate impulse response for each experiment
Y = mean(Ys)   # Average all impulse responses

f1 = plot(vec.(Ys), lab="Individual impulse-response estimates")
plot!(vec(Y), l=(3, :black), lab="Mean")

models = era.(Ys, Ts, 2, 50, 50)    # estimate models based on individual experiments
meanmodel = era(Y, Ts, 2, 50, 50)   # estimate model based on mean impulse response

f2 = bodeplot([G, meanmodel], lab=["True" "" "Combined estimate" ""], l=2)
bodeplot!(models, lab="Individual estimates", c=:black, alpha=0.5, legend=:bottomleft)

plot(f1, f2)
```

![image](https://github.com/baggepinnen/ControlSystemIdentification.jl/assets/3797491/5270a8f3-e441-4549-8cbe-081db9897990)

**Reply by m-bossart** [reply]
- My plan is to formulate the problem such that the initial conditions are always 0.
- The system has one input and one output. The input and output are both complex numbers.

**Reply by m-bossart** [reply]
Thank you for the example, that is very helpful.

**Reply by m-bossart** [reply]
My understanding is that arx is a reasonable starting approach for my system (although maybe not the best)

**Reply by baggepinnen** [reply]
I should warn you that this package has not been tested with complex data, and there are likely places in the code where the adjoint operator `a'` is used instead of `transpose` in a way that would be incorrect for complex data. I looked through the `arx`-related code and didn't see any obvious problems, but be sure to validate your results thoroughly :) 

Since it's a SISO system, `arx` should be a reasonable starting point. `arx` is known to favor a high-frequency fit, but this can be somewhat mitigated by using total least-squares as estimator instead of standard least squares. See the following resources for more info on this
- https://nbviewer.org/github/JuliaControl/ControlExamples.jl/blob/master/identification_arx.ipynb
- https://baggepinnen.github.io/ControlSystemIdentification.jl/dev/tf/#Usage-example:
- https://nbviewer.org/github/JuliaControl/ControlExamples.jl/blob/master/identification_robust.ipynb

**Reply by baggepinnen** [reply]
I made a new release where the support for multiple datasets is now built into `arx` and `era`
